### PR TITLE
Buffer Pool Family

### PR DESCRIPTION
**Pull request by JaiOCP**
For PFC lossless configurations there is a need to have co-related buffer pool configurations across ingress and egress buffer pool objects. Current SAI infra doesn't provide any such co-relation.
A new buffer pool family object is created. 
Ingress and egress buffer pool object will point to same buffer pool family indicating that these are coupled objects.

### CODE REVIEW

**Review by kcudnik — APPROVED** [review]

**Comment by kcudnik** [comment]
approve just for code quality, please discuss on OCP sai meeting

**Comment by rck-innovium** [comment]
@JaiOCP 
Can you please explain the relationships between:

1. SAI_BUFFER_POOL_ATTR_SHARED_SIZE and SAI_BUFFER_POOL_FAMILY_ATTR_SHARED_SIZE 
2. SAI_BUFFER_POOL_ATTR_SIZE and SAI_BUFFER_POOL_FAMILY_ATTR_SIZE

File: inc/saibuffer.h [hunk 1]
```diff
@@ -262,6 +262,48 @@ typedef enum _sai_buffer_pool_threshold_mode_t
...
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_BUFFER_POOL_FAMILY_ATTR_START,
+
+    /**
+     * @brief Shared buffer size in bytes
+     *
+     * This is derived from subtracting all reversed buffers of queue/port
```
Comment (eddyk-nvidia) [inline inc/saibuffer.h:278]: may be reserved buffers ?

**Comment by eddyk-nvidia** [comment]
Could you please clarify the use-case requiring binding of ingress and egress pools ?

**Review by marian-pritsak — CHANGES_REQUESTED** [review]
How will the API user know when to use SAI_BUFFER_POOL_ATTR_SHARED_SIZE and when SAI_BUFFER_POOL_FAMILY_ATTR_SHARED_SIZE?

**Comment by rlhui** [comment]
@JaiOCP - are all comments addressed/resolved?

**Comment by JaiOCP** [comment]
> For PFC lossless configurations there is a need to have co-related buffer pool configurations across ingress and egress buffer pool objects. Current SAI infra doesn't provide any such co-relation.
> A new buffer pool family object is created.
> Ingress and egress buffer pool object will point to same buffer pool family indicating that these are coupled objects.

Its there in the notes of the PR. Recaping it again:
```
For PFC lossless configurations there is a need to have co-related buffer pool configurations across ingress and egress buffer pool objects. Current SAI infra doesn't provide any such co-relation.
A new buffer pool family object is created.
Ingress and egress buffer pool object will point to same buffer pool family indicating that these are coupled objects.
```

File: inc/saibuffer.h (hunk 1 above)
Comment (JaiOCP) [reply inc/saibuffer.h]: https://github.com/opencomputeproject/SAI/pull/1540/commits/91aef8168318af88f0953d22d8e484eb5afd4a45

**Comment by JaiOCP** [comment]
> How will the API user know when to use SAI_BUFFER_POOL_ATTR_SHARED_SIZE and when SAI_BUFFER_POOL_FAMILY_ATTR_SHARED_SIZE?

Added comments to use the query API for attribute value support.
https://github.com/opencomputeproject/SAI/pull/1540/commits/91aef8168318af88f0953d22d8e484eb5afd4a45

**Comment by JaiOCP** [comment]
> @JaiOCP - are all comments addressed/resolved?

All comments are addressed for this PR

**Comment by rlhui** [comment]
@rck-innovium , @marian-pritsak  - would you please help to approve this if no further comments?  Thanks.

**Review by rck-innovium — CHANGES_REQUESTED** [review]

**Comment by rlhui** [comment]
@vmittal-msft - please review? Thanks.

**Pull request closed without merging**